examples/nextjs-with-typescript/src/Link.tsx

```diff
@@ -26,14 +23,17 @@ const NextComposed = React.forwardRef(function NextComposed(
   );
 });
 
-interface LinkProps extends MuiLinkProps, NextLinkProps {
+type LinkProps = LinkPropsBase & Pick<TypographyProps, 'align' | 'color' | 'display' | 'variant'>;
+
+interface LinkPropsBase extends NextLinkProps {
   activeClassName?: string;
+  innerRef?: Ref<HTMLAnchorElement>;
   naked?: boolean;
+  children?: ReactNode;
 }
-
 // A styled version of the Next.js Link component:
 // https://nextjs.org/docs/#with-link
-function RouterLink(props: LinkProps) {
+const RouterLink: FunctionComponent<LinkProps> = props => {
```

Comment:
We never use FunctionComponent in our definitions (NProgressBar exception we should probably change). Is there a reason for changing it here?

Reply:
This is due to tslint [rule](https://palantir.github.io/tslint/rules/only-arrow-functions/) which prefers arrow functions to normal functions. The type of `props` is inferred from the function type.

Wasn't aware of typescript conventions for this project. Is it documented somewhere?

Reply:
Does this rule come by default?

Reply:
Looks like they turn it on for the strictest setting, but the recommended allows named functions too. https://palantir.github.io/tslint/usage/configuration/

Reply:
The above seems to be allowed: https://github.com/palantir/tslint/blob/aa2af999e543f80bcd5f637f8c13316f0ecd5e70/src/configs/recommended.ts#L114.

Reply:
I think that because this is an example, it should support the strictest tslint rules. That way it can be imported into any project with the lowest chance of breaking tslint.

That same recommended file [allows](https://github.com/palantir/tslint/blob/aa2af999e543f80bcd5f637f8c13316f0ecd5e70/src/configs/recommended.ts#L70) the use of `any`, which causes to lose type safety which defeats the purpose of use of TypeScript altogether.

Personally, I've experienced both bugs (people forgetting to use .bind(this)) as well performance hit (if i remember correctly due to some proxy function being created) when using normal functions. Not a big issue either way. I think not having `any` is much more important.

Reply:
>  it should support the strictest tslint rules

I tend to disagree with that. IMO It should compile with `strict` and be a solid starting point for a mui/ts project. I think if they want to use an opinionated coding style like tslint it should up to the user to adapt the example to their needs.

Reply:
>  I think if they want to use an opinionated coding style like tslint it should up to the user to adapt the example to their needs.

Agreed. Lint rules for programming errors or bugs: sure. Stylistic lint rules: no. Following the same convention that create-react-app applies.

This doesn't apply to the `only-arrow-functions` rule since the rationale doesn't apply to function components:

> Traditional functions don’t bind lexical scope, which can lead to unexpected behavior when accessing ‘this’.

Reply:
So what's the rationale for avoiding `FunctionComponent<T>`? Is this for consistency with the rest of the code? To me FunctionComponent is a clear indication that this isn't just some function, but a react component. As [this ](https://www.pluralsight.com/guides/composing-react-components-with-typescript) guide points out, you can have access to the children prop.

It is also being used in one of the best known TypeScript [books](https://basarat.gitbooks.io/typescript/docs/jsx/react.html)

Not an issue reverting back to previous syntax, but just curious, since I thought that was the standard.

Reply:
I see
```jsx
import React, { FunctionComponent, PropsWithChildren } from 'react';

interface Props {
  prop: 'hello';
}

// either:
function Component1 (props: PropsWithChildren<Props>) {
  return <div>{props.children}</div>;
}

// or:
const Component2: FunctionComponent<Props> = props => {
  return <div>{props.children}</div>;
};
```
makes sense to me, but I have no preference.

Reply:
> As this guide points out, you can have access to the children prop.

Implicit children should be avoided since they are neither accurate nor likely to stay in the future.

> So what's the rationale for avoiding FunctionComponent<T>?

I outlined this above. TL;DR: component display names

> To me FunctionComponent is a clear indication that this isn't just some function, but a react component.

99% of the time a uppercase first function is a react component in a react codebase. Uppercase first functions I only ever see in codebases that don't use `class` syntax.